classes/endpoint.php

```diff
@@ -150,4 +150,68 @@ protected function get_oembed_html( $post, $width, $height ) {
 
 		return $output;
 	}
+
+	/**
+	 * Hooks into the REST API output to print XML instead of JSON.
+	 *
+	 * @param bool                      $served  Whether the request has already been served.
+	 * @param WP_HTTP_ResponseInterface $result  Result to send to the client. Usually a WP_REST_Response.
+	 * @param WP_REST_Request           $request Request used to generate the response.
+	 * @param WP_REST_Server            $server  Server instance.
+	 *
+	 * @return bool
+	 */
+	public function rest_pre_serve_request( $served, $result, $request, $server ) {
+		if ( 'HEAD' === $request->get_method() ) {
+			return false;
+		}
+
+		$server->send_header( 'Content-Type', 'application/xml; charset=' . get_option( 'blog_charset' ) );
+
+		// Embed links inside the request.
+		$result = $server->response_to_data( $result, false );
+
+		if ( isset( $result['body'] ) ) {
+			$result = $result['body'];
+		}
+
+		if ( isset( $result[0] ) ) {
+			$result = $result[0];
+		}
+
+		echo '<?xml version="1.0" encoding="' . esc_attr( get_bloginfo( 'charset' ) ) . '" standalone="yes" ?>';
+		echo '<oembed>';
+		foreach ( $result as $key => $value ) {
+			$this->xml_wrap( $key, $value );
+		}
+		echo '</oembed>';
+
+		return $served = true;
+	}
+
+	/**
+	 * Prepare XML output.
+	 *
+	 * @param string $key   XML element name.
+	 * @param mixed  $value XML element content.
+	 */
+	public function xml_wrap( $key, $value ) {
+		if ( is_numeric( $key ) ) {
+			foreach ( $value as $k => $v ) {
+				$this->xml_wrap( $k, $v );
+			}
+
+			return;
+		} else if ( is_array( $value ) ) {
+			echo "<$key>";
+			foreach ( $value as $k => $v ) {
+				$this->xml_wrap( $k, $v );
+			}
+			echo "</$key>";
+
+			return;
+		}
+
+		echo "<$key>" . esc_html( $value ) . "</$key>";
+	}
```

Comment:
Yeah I felt pretty bad when I wrote that… If there isn't already an XML serialiser class available in WordPress or PHP, it seems pretty overkill to build something only for this feature.